File: bindings/mjx-wasm/tests/node/build_a_document.mjs
// The building-a-document walkthrough, written through the WebAssembly binding.
//
// This is `crates/mjx-ooxml/examples/build_a_document.rs` call for call — the same paragraphs, the
// same numbered list, the same hyperlink, the same table, the same header, the same comment, the
// same footnote. `bindings/mjx-python/tests/test_build_a_document.py` is the second copy, and
// `the three Word walkthroughs agree` below checks that this one and the Rust one produce
// byte-identical parts.
//
//     node --test bindings/mjx-wasm/tests/node/
//
// That comparison is the reason this file is shaped the way it is. Until MJXOFF-239 it was not
// here: this file transcribed the Rust example and wrote its own `.docx`, and nothing ever ran the
// Rust one or looked at the two together. Both files passed, which is exactly the failure that is
// invisible from a test report — a `Document` method wired to the wrong facade method produced a
// different document and the suite stayed green, because there was nothing to be different from.
//
// The package under test is the bundler build; `bindings/mjx-wasm/build-npm.sh` produces it.

import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";

import {
  Document,
  Format,
  HeaderFooterType,
  HyperlinkTarget,
  PageSize,
  SectionLocation,
} from "../../npm/dist/bundler/mjx_ooxml.js";

import { partPayloads } from "./zip.mjs";

const HERE = dirname(fileURLToPath(import.meta.url));
const REPOSITORY_ROOT = resolve(HERE, "../../../..");
const OUTPUT_DIRECTORY = process.env.MJX_OUTPUT_DIR ?? join(REPOSITORY_ROOT, "target", "examples");

/**
 * The walkthrough, run once.
 *
 * Everything it builds is freed on the way out, including on failure — which is the shape every
 * caller of this library should copy. A leaked `Document` is a leaked megabyte.
 *
 * @returns {{ saved: Uint8Array, paragraphCount: number }} the document, and what it counted.
 */
function buildTheGuidesDocument() {
  const document = Document.blank(PageSize.a4());
  const owned = [document];
  /** Remembers a wasm object so the `finally` below can free it. */
  const keep = (value) => {
    owned.push(value);
    return value;
  };
  try {
    assert.equal(document.format(), Format.Document);
    assert.equal(document.paragraphCount(), 1);

    // ---- Paragraphs and runs -----------------------------------------------------------------
    document.appendRun(0, "Quarterly Review");
    document.appendParagraph();
    document.appendRun(1, "Prepared by the mjx-ooxml-rs example suite.");
    document.appendParagraph();
    document.appendRun(2, "Highlights");
    document.appendParagraph();
    document.appendRun(3, "Revenue grew across every region this quarter.");

    document.appendParagraph();
    document.appendRun(4, "North America: +12%");
    document.attachParagraphToList(4, 1, 0);
    document.appendParagraph();
    document.appendRun(5, "EMEA: +8%");
    document.attachParagraphToList(5, 1, 0);

    // ---- A hyperlink ---------------------------------------------------------------------------
    document.appendParagraph();
    document.appendRun(6, "Full figures: ");
    const url = keep(HyperlinkTarget.url("https://example.com/investors"));
    document.insertHyperlink(6, 1, "investor relations page", url);

    // ---- A table -------------------------------------------------------------------------------
    const table = document.appendTable(2, 2);
    document.setCellText(table, 0, 0, "Region");
    document.setCellText(table, 0, 1, "Growth");
    document.setCellText(table, 1, 0, "North America");
    document.setCellText(table, 1, 1, "+12%");
    const dimensions = keep(document.tableDimensions(table));
    assert.equal(dimensions.rows, 2);
    assert.equal(dimensions.columns, 2);

    // ---- A header and a comment -----------------------------------------------------------------
    const body = keep(SectionLocation.body());
    document.setHeaderText(body, HeaderFooterType.Default, "Quarterly Review — Internal");
    const commentId = document.addComment(
      0,
      "Reviewer",
      "R",
      "Confirm the North America figure before publishing.",
    );
    assert.notEqual(document.commentRangeText(commentId), undefined);

    // ---- A footnote ----------------------------------------------------------------------------
    document.addFootnote(3, "Figures are unaudited and subject to revision.");

    // ---- Save ----------------------------------------------------------------------------------
    document.validate();
    return { saved: document.save(), paragraphCount: document.paragraphCount() };
  } finally {
    for (const value of owned) {
      value.free();
    }
  }
}

test("the Word walkthrough runs end to end through the wasm binding", () => {
  const run = buildTheGuidesDocument();
  assert.ok(run.saved.length > 0);
  mkdirSync(OUTPUT_DIRECTORY, { recursive: true });
  writeFileSync(join(OUTPUT_DIRECTORY, "wasm_build_a_document.docx"), run.saved);

  // ---- Reopen, to prove the bytes are a real document -----------------------------------------
  const reopened = Document.open(run.saved);
  try {
    assert.equal(reopened.paragraphCount(), run.paragraphCount);
    assert.equal(reopened.paragraphText(0), "Quarterly Review");
    assert.equal(reopened.cellText(0, 0, 0), "Region");
    assert.equal(reopened.headerText(0, HeaderFooterType.Default), "Quarterly Review — Internal");
    assert.equal(reopened.comments().length, 1);
    assert.equal(reopened.footnotes().length, 1);
  } finally {
    reopened.free();
  }
});

test("the three Word walkthroughs agree", (context) => {
  // This walkthrough and the Rust one must produce the *same document*, part for part.
  //
  // Not "both produce a file", and not "both produce a file of about the right size": the same part
  // names, and byte-identical payloads for every one of them. That is the only assertion that can
  // tell a faithful binding from a plausible one — a method wired to the wrong `Document` method, a
  // paragraph index off by one, or an argument converted with the wrong units, changes a payload
  // here and nothing else would notice.
  let rustOutput;
  try {
    mkdirSync(OUTPUT_DIRECTORY, { recursive: true });
    rustOutput = join(OUTPUT_DIRECTORY, "facade_build_a_document.docx");
    execFileSync(
      "cargo",
      ["run", "--quiet", "-p", "mjx-ooxml", "--example", "build_a_document", "--", rustOutput],
      { cwd: REPOSITORY_ROOT, stdio: "pipe" },
    );
  } catch (failure) {
    context.skip(`the Rust walkthrough could not be run: ${failure.message.split("\n")[0]}`);
    return;
  }

  const fromNode = partPayloads(buildTheGuidesDocument().saved);
  const fromRust = partPayloads(readFileSync(rustOutput));

  assert.deepEqual(
    Object.keys(fromNode).sort(),
    Object.keys(fromRust).sort(),
    "the two walkthroughs must author the same set of parts",
  );
  const differing = Object.keys(fromNode).filter(
    (name) => Buffer.compare(fromNode[name], fromRust[name]) !== 0,
  );
  assert.deepEqual(
    differing,
    [],
    `these parts differ between the Node and Rust walkthroughs: ${differing.join(", ")}`,
  );
});

File: crates/mjx-ooxml/examples/build_a_document.rs
//! The building-a-document walkthrough, written entirely through the facade.
//!
//! `bindings/mjx-python/tests/test_build_a_document.py` and
//! `bindings/mjx-wasm/tests/node/build_a_document.mjs` are the same walkthrough, call for call, in
//! the other two languages MJXOFF-139 curated `Document` for — proof that the curated surface is
//! actually enough to author a real document, not merely to open one.
//!
//! ```sh
//! cargo run -p mjx-ooxml --example build_a_document -- out.docx
//! ```
//!
//! Note where the file I/O is: right here, in the caller. The library is bytes-in and bytes-out and
//! never touches a filesystem — which is exactly why the same calls work unchanged in a browser.

use std::path::PathBuf;

use mjx_ooxml::{Document, Format, HeaderFooterType, HyperlinkTarget, PageSize, SectionLocation};

/// Where this example writes: its first argument, or `target/examples/` by default.
fn output_path() -> PathBuf {
    match std::env::args().nth(1) {
        Some(path) => PathBuf::from(path),
        None => {
            let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../target/examples");
            let _ = std::fs::create_dir_all(&dir);
            dir.join("facade_build_a_document.docx")
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let out = output_path();

    // ---- Blank -------------------------------------------------------------------------------
    // Nothing is read from disk: every part is authored from this library's own element builders,
    // which is what makes a document buildable from a `pip install` or a browser tab with no input
    // file.
    let mut document = Document::blank(PageSize::a4())?;
    assert_eq!(document.format(), Format::Document);
    assert_eq!(document.paragraph_count()?, 1);

    // ---- Paragraphs and runs ------------------------------------------------------------------
    // `Document::blank` writes one *empty* paragraph — no run to edit yet, so the first line is
    // authored with `append_run` rather than `set_run_text`.
    document.append_run(0, "Quarterly Review")?;
    document.append_paragraph()?;
    document.append_run(1, "Prepared by the mjx-ooxml-rs example suite.")?;
    document.append_paragraph()?;
    document.append_run(2, "Highlights")?;
    document.append_paragraph()?;
    document.append_run(3, "Revenue grew across every region this quarter.")?;

    // A numbered list: two more paragraphs, attached to numbering instance 1 at level 0. (No
    // `word/numbering.xml` exists yet on a blank document; `attach_paragraph_to_list` writes the
    // reference regardless — resolving it against a real list definition is a step a caller adds
    // through `Document::document_mut` when one exists.)
    document.append_paragraph()?;
    document.append_run(4, "North America: +12%")?;
    document.attach_paragraph_to_list(4, 1, 0)?;
    document.append_paragraph()?;
    document.append_run(5, "EMEA: +8%")?;
    document.attach_paragraph_to_list(5, 1, 0)?;

    // ---- A hyperlink ---------------------------------------------------------------------------
    document.append_paragraph()?;
    document.append_run(6, "Full figures: ")?;
    document.insert_hyperlink(
        6,
        1,
        "investor relations page",
        &HyperlinkTarget::Url("https://example.com/investors".to_owned()),
    )?;

    // ---- A table ---------------------------------------------------------------------------------
    let table = document.append_table(2, 2)?;
    document.set_cell_text(table, 0, 0, "Region")?;
    document.set_cell_text(table, 0, 1, "Growth")?;
    document.set_cell_text(table, 1, 0, "North America")?;
    document.set_cell_text(table, 1, 1, "+12%")?;
    let (rows, columns) = document.table_dimensions(table)?;
    assert_eq!((rows, columns), (2, 2));

    // ---- A header and a comment -------------------------------------------------------------------
    document.set_header_text(
        SectionLocation::Body,
        HeaderFooterType::Default,
        "Quarterly Review — Internal",
    )?;
    let comment_id = document.add_comment(
        0,
        "Reviewer",
        Some("R"),
        "Confirm the North America figure before publishing.",
    )?;
    assert!(document.comment_range_text(comment_id)?.is_some());

    // ---- A footnote --------------------------------------------------------------------------------
    document.add_footnote(3, "Figures are unaudited and subject to revision.")?;

    // ---- Save --------------------------------------------------------------------------------------
    document.validate()?;
    let bytes = document.save()?;
    std::fs::write(&out, &bytes)?;
    println!("wrote {} bytes to {}", bytes.len(), out.display());

    // ---- Reopen, to prove the bytes are a real document --------------------------------------------
    let mut reopened = Document::open(&bytes)?;
    assert_eq!(reopened.paragraph_count()?, document.paragraph_count()?);
    assert_eq!(reopened.paragraph_text(0)?, "Quarterly Review");
    assert_eq!(reopened.cell_text(0, 0, 0)?, "Region");
    assert_eq!(
        reopened
            .header_text(0, HeaderFooterType::Default)?
            .as_deref(),
        Some("Quarterly Review — Internal")
    );
    assert_eq!(reopened.comments()?.len(), 1);
    assert_eq!(reopened.footnotes()?.len(), 1);

    Ok(())
}
